x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -0,0 +1,541 @@
+            }
+            return new ConstantScoreQuery(bq.build());
+        }     
+                
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
+            failIfNoDocValues();
+            return new DocValuesIndexFieldData.BinaryBuilder();
+        }
+        
+    }
+
+    private int ignoreAbove;
+    private int numChars;
+
+    private WildcardFieldMapper(String simpleName, MappedFieldType fieldType, MappedFieldType defaultFieldType,
+                int ignoreAbove, int numChars, Settings indexSettings, MultiFields multiFields, CopyTo copyTo) {
+        super(simpleName, fieldType, defaultFieldType, indexSettings, multiFields, copyTo);
+        this.ignoreAbove = ignoreAbove;
+        this.numChars = numChars;
+        assert fieldType.indexOptions() == IndexOptions.DOCS;
+    }
+
+
+    /** Values that have more chars than the return value of this method will
+     *  be skipped at parsing time. */
+    // pkg-private for testing
+    int ignoreAbove() {
+        return ignoreAbove;
+    }    
+    
+    @Override
+    protected WildcardFieldMapper clone() {
+        return (WildcardFieldMapper) super.clone();
+    }
+
+    @Override
+    public WildcardFieldType fieldType() {
+        return (WildcardFieldType) super.fieldType();
+    }
+
+    @Override
+    protected void doXContentBody(XContentBuilder builder, boolean includeDefaults, Params params) throws IOException {
+        super.doXContentBody(builder, includeDefaults, params);
+        builder.field("num_chars", fieldType().numChars());
+        if (includeDefaults || ignoreAbove != Defaults.IGNORE_ABOVE) {
+            builder.field("ignore_above", ignoreAbove);
+        }
+    }
+    
+    @Override
+    protected void parseCreateField(ParseContext context, List<IndexableField> fields) throws IOException {
+        final String value;
+        if (context.externalValueSet()) {
+            value = context.externalValue().toString();
+        } else {
+            XContentParser parser = context.parser();
+            if (parser.currentToken() == XContentParser.Token.VALUE_NULL) {
+                value = fieldType().nullValueAsString();
+            } else {
+                value =  parser.textOrNull();
+            }
+        }
+        createFields(value, fields);        
+    }   
+    
+    void createFields(String value, List<IndexableField>fields) {
+        if (value == null || value.length() > ignoreAbove) {
+            return;
+        }
+        TaperedNgramTokenizer tokenizer = new TaperedNgramTokenizer(fieldType().numChars);
+        tokenizer.setReader(new StringReader(TOKEN_START_OR_END_CHAR+ value +TOKEN_START_OR_END_CHAR));
+        
+        Field field = new Field(fieldType().name(), tokenizer, fieldType());
+        fields.add(field);
+        
+        Field dvField = new BinaryDocValuesField(fieldType().name(), new BytesRef(value));        
```

Comment:
Unfortunately, this will only work for single-valued fields since `BinaryDocValuesField` supports a single value per document. Can you use similar logic as `BinaryFieldMapper` to support multi-valued fields?